internal/pkg/agent/storage/store/internal/migrations/migrations.go

```diff
@@ -0,0 +1,96 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package migrations
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"gopkg.in/yaml.v2"
+)
+
+type StateStore struct {
+	Action      Action        `yaml:"action"`
+	AckToken    string        `yaml:"ack_token"`
+	ActionQueue []ActionQueue `yaml:"action_queue"`
+}
+
+// Action is a struct to read an Action Policy Change from its old
+// YAML format.
+type Action struct {
+	ActionID     string                 `yaml:"action_id"`
+	Type         string                 `yaml:"action_type"`
+	StartTime    time.Time              `yaml:"start_time,omitempty"`
+	SourceURI    string                 `yaml:"source_uri,omitempty"`
+	RetryAttempt int                    `yaml:"retry_attempt,omitempty"`
+	Policy       map[string]interface{} `yaml:"policy,omitempty"`
+	IsDetected   bool                   `yaml:"is_detected,omitempty"`
+}
+
+// ActionQueue is a struct to read the action queue from its old YAML format.
+type ActionQueue struct {
+	ActionID       string                 `yaml:"action_id"`
+	Type           string                 `yaml:"type"`
+	StartTime      time.Time              `yaml:"start_time,omitempty"`
+	ExpirationTime time.Time              `yaml:"expiration,omitempty"`
+	Version        string                 `yaml:"version,omitempty"`
+	SourceURI      string                 `yaml:"source_uri,omitempty"`
+	RetryAttempt   int                    `yaml:"retry_attempt,omitempty"`
+	Policy         map[string]interface{} `yaml:"policy,omitempty"`
+	IsDetected     bool                   `yaml:"is_detected,omitempty"`
+}
+
+type loader interface {
+	Load() (io.ReadCloser, error)
+}
+
+// LoadActionStore loads an action store from loader.
+func LoadActionStore(loader loader) (*Action, error) {
+	return LoadStore[Action](loader)
+}
+
+// LoadYAMLStateStore loads the old YAML state store from loader.
+func LoadYAMLStateStore(loader loader) (*StateStore, error) {
+	return LoadStore[StateStore](loader)
+}
+
+// LoadStore loads a YAML file.
+func LoadStore[Store any](loader loader) (store *Store, err error) {
+	// Store is a generic type, this might be needed.
+	store = new(Store)
+
+	reader, err := loader.Load()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load action store: %w", err)
+	}
+	defer func() {
+		errClose := reader.Close()
+		if errClose != nil {
+			errClose = fmt.Errorf(
+				"migration storeLoad failed to close reader: %w", errClose)
+		}
+		err = errors.Join(err, errClose)
+	}()
```

Comment:
@cmacknz, answering your question:

I tested on windows and as I expected it breaks if it fails to close the reader.
Also the current implementation only returns the error if there is already an error. But I changed it now, better to return it anyway.

here is what happens if the reader isn't closed:
```
--- FAIL: TestStoreMigrations (1.65s)
    --- FAIL: TestStoreMigrations/action_store_to_YAML_state_store (1.65s)
        --- FAIL: TestStoreMigrations/action_store_to_YAML_state_store/unenroll (1.65s)
            migrations_test.go:147:
                        Error Trace:    C:/vagrant/internal/pkg/agent/storage/store/migrations_test.go:147
                        Error:          Received unexpected error:
                                        migration storeLoad failed to close reader: fake reader.Close() error
                        Test:           TestStoreMigrations/action_store_to_YAML_state_store/unenroll
                        Messages:       migration action store -> state store failed
            testing.go:1225: TempDir RemoveAll cleanup: remove C:\Users\vagrant\AppData\Local\Temp\1\TestStoreMigrationsaction_store_to_YAML_state_storeunenroll2169855176\001\action_store.yml: The process cannot access the file because it is being used by another process.
FAIL
FAIL    github.com/elastic/elastic-agent/internal/pkg/a
```

to emulate this error, I just removed the code closing the reader